### PR TITLE
Fix lasagne updates

### LINKED ISSUE #402: lasagne.updates.adam yields float64 updates with float32 0d-array input
Just tracked down this strange behavior.

```Python
>>> import theano
>>> import lasagne
>>> import numpy as np
>>> import theano.tensor as T
>>> x = theano.shared(np.ones((10,), dtype=theano.config.floatX))
>>> f = T.sum(x**2)
>>> u = lasagne.updates.adam(f, [x], beta2=.99)
>>> print [v.type for v in u.values()]
[TensorType(float32, vector),
 TensorType(float32, vector),
 TensorType(float32, vector),
 TensorType(float32, scalar)]
>>> u = lasagne.updates.adam(
...     f, [x], beta2=np.array(.99, dtype=theano.config.floatX))
>>> print [v.type for v in u.values()]
[TensorType(float32, vector),
 TensorType(float64, vector),
 TensorType(float64, vector),
 TensorType(float32, scalar)]
```

If the resulting updates are passed into `theano.function`, a type error is raised because the update for a `float32` values is `float64`.  In this particular case, what's happening is that in `lasagne.adams.updates`, the following:
```
v_t = beta2*v_prev + (1-beta2)*g_t**2
```
`(1 - beta2)` will be given a dtype `float64`:
```Python
>>> (1 - np.array(0., dtype='float32')).dtype
dtype('float64')
```

So, given this, I think we have three options:

1. Implicitly cast things using `utils.floatX` everywhere, which would make the line in question become
`beta2*v_prev + (utils.floatX(1.)-beta2)*g_t**utils.floatX(2.)` (puke)

1. Convert arguments to updates functions to `float`, because if a float is passed instead of a 0d-array with the correct dtype, everything works out fine thanks to Theano's willingness to cast Python scalars to the correct type

1. Tell users (like me) to RTFM, i.e. `beta2 : float` is specified in the docstring, so if I pass anything else (like a numpy scalar), I'm going to have a bad time.

### LINKED ISSUE #563: dropout layers leak float64 into graph
When using a dropout layer, the output upcasts to float64, causing every layer after it to be upcast to float64. The problem seems to be caused by the way numpy treats operations that upcast integers. For example:

x = 1 - numpy.float32(0.5)

causes x to have type numpy.float64

This happens to me on numpy version 1.9.0, but I think it happens on other versions as well.

So, here is a code fix:

Line 73 of noise.py - which reads:
retain_prob = 1 - self.p

should be changed to something like this:
if theano.config.floatX == 'float64' :
   retain_prob = np.float64(1 - self.p)
else :
   retain_prob = np.float32(1 - self.p)

### PR DESCRIPTION
**Pull request by sarin1991**
These changes fix the upcasting issue in lasagne.update functions rmsprop, adadelta, adam and adamax where float32 inputs are given float64 updates 
A test method, test_update_returntype has also been added to the class TestUpdateFunctions which checks for this error.
Closes #402. Closes #563. Closes #578.

### CODE REVIEW

**Comment by Seanny123** [comment]
LGTM.

**Comment by benanne** [comment]
Awesome work, thanks a lot! Just two remarks, both fairly minor:

- I wonder if there's some way we can make this less verbose, all the `as_theano_expression` calls add quite a lot of cruft. Maybe it could be as simple as creating a simple multi-argument shortcut function so we could do something like `learning_rate, rho, epsilon = no_upcast(learning_rate, rho, epsilon)`.

- When adding a new updates function, the coverage checker will not tell us that we also need to add some parameters to this test as well. I wonder if there's a way to make things more general, so this is less likely to be forgotten.

Just some thoughts, I would be fine with seeing this merged as-is though. LGTM :)

File: lasagne/tests/test_updates.py [hunk 1]
```diff
@@ -65,6 +65,53 @@ def test_updates(self, method, kwargs):
         assert np.allclose(A.get_value(), B.get_value())
         assert np.allclose(A.get_value(), self.torch_values[method])
 
+    @pytest.mark.parametrize('method, kwargs', [
+        ['sgd', {'learning_rate': 0.1}],
+        ['sgd', {'learning_rate': np.float32(0.1)}],
```
Comment (f0k) [inline lasagne/tests/test_updates.py]: 1. Instead of listing everything twice, can you run it once with floats and then with np.float32 in the test function body?
2. Instead of listing all the arguments, would it be possible to just list the methods and then in the function body, use `inspect.getargspec` to get the default arguments and once call with the defaults and then with the defaults converted to np.float32? This would make it a lot easier to extend if more functions are added.

File: lasagne/updates.py [hunk 2]
```diff
@@ -440,6 +440,12 @@ def rmsprop(loss_or_grads, params, learning_rate=1.0, rho=0.9, epsilon=1e-6):
            Neural Networks for Machine Learning, Lecture 6.5 - rmsprop.
            Coursera. http://www.youtube.com/watch?v=O3sxAc4hxZU (formula @5:20)
     """
+    # Converting input arguments to theano expressions so that
+    # float32 inputs arent upcasted to float 64 automatically
+    learning_rate = utils.as_theano_expression(learning_rate)
+    rho = utils.as_theano_expression(rho)
+    epsilon = utils.as_theano_expression(epsilon)
```
Comment (f0k) [inline lasagne/updates.py]: I think I'd prefer only converting what's needed. Something like:
```python
# Convert rho to theano expression so (1 - rho) will never be casted to float64
rho = utils.as_theano_expression(rho)
```
The others don't risk upcasting anyway. Similar for adam and adamax: only beta1, beta2 need to be converted.

**Comment by sarin1991** [comment]
@benanne  I have added in a function no_upcast to utils pretty much along the lines you suggested. I was initially thinking of doing the same but wasn't sure if you guys were okay with me adding another function to utils.py. I probably need to improve the documentation. What do you think?

@f0k I can remove the variables which are not necessary if you think that's better but I was thinking we should do this check for all the arguments including the ones sent to layer instances. As it can cause weird errors in the future if the people developing it are not careful about this upcasting.

File: lasagne/utils.py [hunk 3]
```diff
@@ -21,6 +21,29 @@ def floatX(arr):
     return np.asarray(arr, dtype=theano.config.floatX)
 
 
+def no_upcast(list_element):
```
Comment (craffel) [inline lasagne/utils.py]: Why not use `*args`?  This would save you from having to construct a list to call the function, which would make it cleaner, IMO:
```Python
def no_upcast(*args):
    result_list = []
     for element in args:
         if isinstance(element, np.ndarray):
             result_list.append(T.constant(element))
         else:
             result_list.append(element)
     return result_list
```
and the call signature would be e.g.
```
learning_rate, rho, epsilon = utils.no_upcast(
    learning_rate, rho, epsilon)
```

... or, better yet, make it a pretty one-liner with a list comprehension
```
def no_upcast(*args):
     return (T.constant(e) if isinstance(e, np.ndarray) else e for e in args)
```

**Comment by craffel** [comment]
Thanks for doing this!  We need it.  Just made a comment re: your `no_upcast` function, I personally think it would be cleaner to just use `*args` rather than forcing an iterable to be passed in.  The function could also be written in a readable one-liner.  Otherwise LGTM after a cursory glance.

**Comment by f0k** [comment]
> I have added in a function no_upcast to utils pretty much along the lines you suggested. I was initially thinking of doing the same but wasn't sure if you guys were okay with me adding another function to utils.py. I probably need to improve the documentation. What do you think?

To be honest, I liked it better before. Firstly, `no_upcast` is not a suitable name because it still only wraps things as Theano variables, we just happen to need it in one case to sidestep numpy's upcasting rules. And secondly, I don't see a need for an extra function just to save two or three lines in total.

> I can remove the variables which are not necessary if you think that's better but I was thinking we should do this check for all the arguments including the ones sent to layer instances.

Fair enough, but you're not doing it for all the arguments either, just for the update functions that were affected. And I wouldn't want to expand this to all attributes of all layers, we couldn't do any comparisons to constants then:
```
>>> 1 == 1
True
>>> np.array(1) == 1
True
>>> T.constant(1) == 1
False
```

> As it can cause weird errors in the future if the people developing it are not careful about this upcasting.

Seeing where this leads, isn't it easier to just wrap the few Python constants we have in `utils.floatX()`? People need to be careful in either case.

**Comment by sarin1991** [comment]
@craffel Yes, using *args looks a lot cleaner. I will implement your suggestion if we end up using a function in utils. Thanks! :)

@f0k `utils.floatX()` doesn't work. It breaks when you give python float arguments.
```
>>>from lasagne import utils
>>>beta1 = 0.99
>>>one = utils.floatX(1)
>>>one.dtype
dtype('float32')
>>>(one-beta1).dtype
dtype('float64')
```

> Fair enough, but you're not doing it for all the arguments either, just for the update functions that were affected. And I wouldn't want to expand this to all attributes of all layers, we couldn't do any comparisons to constants

I didn't realize that we were doing comparison operations since I thought all the parameters could be replaced as a theano expression. You are right that would be a problem. We could use theano's get_scalar_constant_value function though but I guess it might be easier to just use `utils.floatX()` for dropout layer.

In any case, I don't mind whether we use a new function or not. I will wait for a bit to see what approach is acceptable by everyone here before making any further changes :)

**Comment by f0k** [comment]
> `utils.floatX()` doesn't work. It breaks when you give python float arguments.

Bummer... now that's a good argument against it :) Guess I only checked the beta1=ndarray case.

> I didn't realize that we were doing comparison operations since I thought all the parameters could be replaced as a theano expression.

We'd need to check in detail. Some parameters will accept Theano expressions without problems, some with minor disadvantages (e.g., the dropout rate can be a Theano expression, but then it might not be optimized away if it's set to zero), some won't (e.g., the number of units in dense layers). What about turning the `1` into a Theano expression instead of the `beta` parameters?
```
>>> one = T.constant(1)
>>> (one - .99).dtype
'float32'
>>> (one - np.array(.99, dtype=np.float32)).dtype
'float32'
>>> (one - T.scalar()).dtype
'float32'
```
This would cover all cases and be less invasive. What do you think? What do others think?

 `T.constant(1.)` and `T.constant(1)` seem to give the same compiled graph (e.g., `T.matrix() - T.constant(1)` will be executed on GPU with `T.constant(1)` becoming a float32 CUDA array even if `T.constant(1)` is `int8` originally).

**Comment by sarin1991** [comment]
That's a great idea! :) Yes, I agree it should work for all the cases and is less invasive

**Comment by sarin1991** [comment]
I have updated the fix using T.constant in the expression. I have also added the fix for the Dropout Layer and added a unit test for it

File: lasagne/tests/layers/test_noise.py [hunk 4]
```diff
@@ -60,6 +65,11 @@ def test_get_output_for_p_02(self, layer_p_02):
         assert 0.9 < result_eval.mean() < 1.1
         assert (numpy.round(numpy.unique(result_eval), 2) == [0., 1.25]).all()
 
+    def test_get_output_for_p_float32_dtype(self, layer_p_float32):
+        input = theano.shared(numpy.ones((100, 100), dtype=numpy.float32))
+        result = layer_p_float32.get_output_for(input)
+        assert result.dtype == input.dtype
```
Comment (f0k) [inline lasagne/tests/layers/test_noise.py]: As `layer_p_float32` is only used once, I'd prefer not to define it and write the test directly as:
```python
def test_get_output_for_float32_p(self):
    floatX_ = theano.config.floatX
    theano.config.floatX = 'float32'
    from lasagne.layers.noise import DropoutLayer
    layer = DropoutLayer((None, 100), p=numpy.float32(0.5))
    input = T.matrix()
    assert layer.get_ouput_for(input).dtype == input.dtype
    theano.config.floatX = floatX_
```

File: lasagne/updates.py [hunk 5]
```diff
@@ -440,14 +440,17 @@ def rmsprop(loss_or_grads, params, learning_rate=1.0, rho=0.9, epsilon=1e-6):
            Neural Networks for Machine Learning, Lecture 6.5 - rmsprop.
            Coursera. http://www.youtube.com/watch?v=O3sxAc4hxZU (formula @5:20)
     """
+    # Using theano constant to prevent upcasting of float32
+    one = T.constant(1)
+
```
Comment (f0k) [inline lasagne/updates.py]: Nitpick, but I'd prefer to move this *after* `grads = ` and `updates = ` so all update functions start the same.

**Comment by sarin1991** [comment]
@f0k What do you think? Everything okay now?

File: lasagne/updates.py [hunk 6]
```diff
@@ -440,14 +440,18 @@ def rmsprop(loss_or_grads, params, learning_rate=1.0, rho=0.9, epsilon=1e-6):
            Neural Networks for Machine Learning, Lecture 6.5 - rmsprop.
            Coursera. http://www.youtube.com/watch?v=O3sxAc4hxZU (formula @5:20)
     """
+
```
Comment (f0k) [inline lasagne/updates.py]: Please remove this extra whitespace.

File: lasagne/updates.py [hunk 7]
```diff
@@ -508,9 +512,13 @@ def adadelta(loss_or_grads, params, learning_rate=1.0, rho=0.95, epsilon=1e-6):
            ADADELTA: An Adaptive Learning Rate Method.
            arXiv Preprint arXiv:1212.5701.
     """
+
```
Comment (f0k) [inline lasagne/updates.py]: And this.

File: lasagne/updates.py [hunk 8]
```diff
@@ -574,12 +582,16 @@ def adam(loss_or_grads, params, learning_rate=0.001, beta1=0.9,
            Adam: A Method for Stochastic Optimization.
            arXiv preprint arXiv:1412.6980.
     """
+
```
Comment (f0k) [inline lasagne/updates.py]: And this.

File: lasagne/updates.py [hunk 9]
```diff
@@ -633,12 +645,16 @@ def adamax(loss_or_grads, params, learning_rate=0.002, beta1=0.9,
...
```
Comment (f0k) [inline lasagne/updates.py]: And this.

File: lasagne/tests/test_updates.py [hunk 10]
```diff
@@ -65,6 +65,52 @@ def test_updates(self, method, kwargs):
...
+
+        # Checking for float32 arguments
+        for param in kwargs:
+            kwargs[param] = np.float32(kwargs[param])
+        updates = update_func(self.f(A) + self.f(B),
+                              [A, B],
+                              **kwargs)
+
+        assert all(v.dtype == 'float32' for v in updates)
+        theano.config.floatX = floatX_
```
Comment (f0k) [inline lasagne/tests/test_updates.py]: Very good, thank you! Just one last thing: We should probably wrap this in a try..finally block, otherwise all other tests will be affected if this one fails:
```python
floatX_ = theano.config.floatX
theano.config.floatX = 'float32'
try:
    A = ...
    B = ...
    ...
    assert all(...)
finally:
    theano.config.floatX = floatX_
```
Sorry, I didn't think of this before.

**Comment by sarin1991** [comment]
@f0k How about now?

**Comment by f0k** [comment]
Perfect! Now we just need to reorganize the commits a bit. Can you do:
```bash
git reset 0440814  # that was were you branched off
git add lasagne/updates.py
git add lasagne/tests/test_updates.py
git commit -m 'Avoid upcast to float64 in lasagne.updates'
git add lasagne/layers/noise.py
git add lasagne/tests/layers/test_noise.py
git commit -m 'Avoid upcast to float64 in DropoutLayer'
git push --force
```
This will group all your changes into two coherent commits. Let me know when you're done and I can merge!

**Comment by sarin1991** [comment]
@f0k Done!

**Comment by f0k** [comment]
Yay, thanks a lot for all your work! Merging!

@benanne: coveralls is back, did you do anything or did that happen on its own?

**Comment by sarin1991** [comment]
Sure no problem! 👍

**Comment by benanne** [comment]
@f0k not sure, I think I "regenerated the repository token" a couple of days ago, whatever that means. Maybe that did it. Maybe not :)
